ravenframework/Models/ROM.py

```diff
@@ -567,3 +598,124 @@ def writeXML(self, what='all'):
+      @ Out, xml, xmlUtils.StaticXmlElement, written meta
+    """
+    template = r"""
+# MODEL GENERATED BY RAVEN (raven.inl.gov)
+import pyomo.environ as pyo
+from pyomo.contrib.pynumero.interfaces.external_grey_box import ExternalGreyBoxModel, ExternalGreyBoxBlock
+from pyomo.contrib.pynumero.dependencies import (numpy as np)
+from pyomo.contrib.pynumero.asl import AmplInterface
+from pyomo.contrib.pynumero.algorithms.solvers.cyipopt_solver import CyIpoptSolver, CyIpoptNLP
+import os, sys, pickle
+# RAVEN ROM PYOMO GREY MODEL CLASS
+
+class ravenROM(ExternalGreyBoxModel):
+
+  def __init__(self, **kwargs):
+    self._rom_file = kwargs.get("rom_file")
+    self._raven_framework = kwargs.get("raven_framework")
+    self._raven_framework = os.path.abspath(self._raven_framework)
+    if not os.path.exists(self._raven_framework):
+      raise IOError('The RAVEN framework directory does not exist in location "' + str(self._raven_framework)+'" !')
+    if os.path.isdir(os.path.join(self._raven_framework,"ravenframework")):
+      sys.path.append(self._raven_framework)
+    else:
+      raise IOError('The RAVEN framework directory does not exist in location "' + str(self._raven_framework)+'" !')
+    from ravenframework.CustomDrivers import DriverUtils as dutils
+    dutils.doSetup()
+    # de-serialize the ROM
+    self._rom_file = os.path.abspath(self._rom_file)
+    if not os.path.exists(self._rom_file):
+      raise IOError('The serialized (binary) file has not been found in location "' + str(self._rom_file)+'" !')
+    self.rom = pickle.load(open(self._rom_file, mode='rb'))
+    #
+    self.settings = self.rom.getInitParams()
+    # get input names
+    self._input_names = self.settings.get('Features')
+    # n_inputs
+    self._n_inputs = len(self._input_names)
+    # get output names
+    self._output_names = self.settings.get('Target')
+    # n_outputs
+    self._n_outputs = len(self._output_names)
+    # input values storage
+    self._input_values = np.zeros(self._n_inputs, dtype=np.float64)
+
+  def return_train_values(self, feat):
+    return self.rom.trainingSet.get(feat)
+
+  def input_names(self):
+    return self._input_names
+
+  def output_names(self):
+    return self._output_names
+
+  def set_input_values(self, input_values):
+    assert len(input_values) == self._n_inputs
+    np.copyto(self._input_values, input_values)
+
+  def evaluate_equality_constraints(self):
+      raise NotImplementedError('This method should not be called for this model.')
+
+  def evaluate_outputs(self):
+    request = {k:np.asarray(v) for k,v in zip(self._input_names,self._input_values)}
+    outs = self.rom.evaluate(request)
+    eval_outputs = np.asarray([outs[k].flatten() for k in self._output_names], dtype=np.float64)
+    return eval_outputs.flatten()
+
+  def evaluate_jacobian_outputs(self):
+    request = {k:np.asarray(v) for k,v in zip(self._input_names,self._input_values)}
+    derivatives = self.rom.derivatives(request)
```

Comment:
Could you also add controls for "feats" and "orders" in rom.derivatives?

Reply:
I can sure check if I can. However, since it is a printed code, the current inout structure does not allow the user to choose the orders of the derivative.

Reply:
Examining the code structure, "feats" would be difficult to add since the user must know what the features are from a ROM that they potentially did not run assuming they are not developers.

However, I think it is appropriate to add "order". The user can test which order optimized the problem the best. This capability has been added.